### PR TITLE
improve subscription pricing behavior

### PR DESCRIPTION
**Pull request by sqs**
- Fix bugs
- Support subscriptions that do not require payment (free plan)
- Support min and max user counts for pricing plans

### CODE REVIEW

File: web/src/enterprise/user/productSubscriptions/ProductSubscriptionForm.tsx [hunk 1]
```diff
@@ -233,8 +251,10 @@ const _ProductSubscriptionForm: React.FunctionComponent<Props & ReactStripeEleme
                                     <>
                                         <LoadingSpinner className="icon-inline mr-2" /> Processing...
                                     </>
-                                ) : (
+                                ) : paymentValidity !== PaymentValidity.NoPaymentRequired ? (
                                     primaryButtonText
```
Comment (felixfbecker) [inline web/src/enterprise/user/productSubscriptions/ProductSubscriptionForm.tsx:255]: This adds another case to the conditional, but this doesn't seem to be hit in any test case

File: web/src/enterprise/dotcom/productPlans/ProductPlanPrice.tsx [hunk 2]
```diff
@@ -1,13 +1,27 @@
 import React from 'react'
 import * as GQL from '../../../../../shared/src/graphql/schema'
+import { ProductPlanTiered } from './ProductPlanTiered'
+import { pluralize } from '../../../../../shared/src/util/strings'
 
 /** Displays the price of a plan. */
 export const ProductPlanPrice: React.FunctionComponent<{
```
Comment (felixfbecker) [inline web/src/enterprise/dotcom/productPlans/ProductPlanPrice.tsx:7]: This looks like a candidate for an easy snapshot test

File: web/src/enterprise/user/productSubscriptions/ProductSubscriptionForm.tsx [hunk 3]
```diff
@@ -115,14 +128,14 @@ const _ProductSubscriptionForm: React.FunctionComponent<Props & ReactStripeEleme
                 submits.pipe(
                     switchMap(() =>
                         // TODO(sqs): store name, address, company, etc., in token
-                        from(stripe.createToken()).pipe(
+                        (paymentValidity !== PaymentValidity.NoPaymentRequired
```
[the remark this answers is not in the document]
Comment (sqs) [reply web/src/enterprise/user/productSubscriptions/ProductSubscriptionForm.tsx:131]: Yeah. I created https://github.com/sourcegraph/sourcegraph/issues/10366 to follow up here.

**Comment by sqs** [comment]
Thanks. All comments addressed, one followup created: https://github.com/sourcegraph/sourcegraph/issues/10366.

**Comment by sqs** [comment]
@unknwon Sorry, I missed your comment for some reason.

> Go code LGTM, but I notice all changed Go files (except schema.go) has 0% code coverage, can we add some?

Yes, I added this to https://github.com/sourcegraph/sourcegraph/issues/10366 and will work on this in the next few weeks.

**Pull request merged**